icu4c/source/common/uresbund.cpp

```diff
@@ -91,6 +91,202 @@ static UBool chopLocale(char *name) {
     return FALSE;
 }
 
+static UBool hasVariant(const char* localeID) {
+    UErrorCode err = U_ZERO_ERROR;
+    int32_t variantLength = uloc_getVariant(localeID, NULL, 0, &err);
+    return variantLength != 0;
+}
+
+// This file contains the tables for doing locale fallback, which are generated
+// by the CLDR-to-ICU process directly from the CLDR data.  This file should only
+// ever be included from here.
+#define INCLUDED_FROM_URESBUND_CPP
+#include "localefallback_data.h"
+
+static const char* performFallbackLookup(const char* key,
+                                         const char* keyStrs,
+                                         const char* valueStrs,
+                                         const int32_t* lookupTable,
+                                         int32_t lookupTableLength) {
+    const int32_t* bottom = lookupTable;
+    const int32_t* top = lookupTable + lookupTableLength;
+
+    while (bottom < top) {
+        // Effectively, divide by 2 and round down to an even index
+        const int32_t* middle = bottom + (((top - bottom) / 4) * 2);
+        const char* entryKey = &(keyStrs[*middle]);
+        int32_t strcmpResult = uprv_strcmp(key, entryKey);
+        if (strcmpResult == 0) {
+            return &(valueStrs[middle[1]]);
+        } else if (strcmpResult < 0) {
+            top = middle;
+        } else {
+            bottom = middle + 2;
+        }
+    }
+    return nullptr;
+}
+
+static CharString getDefaultScript(const CharString& language, const CharString& region) {
+    const char* defaultScript = nullptr;
+    UErrorCode err = U_ZERO_ERROR;
+    
+    // if we were passed both language and region, make them into a locale ID and look that up in the default
+    // script table
+    if (!region.isEmpty()) {
+        CharString localeID;
+        localeID.append(language, err).append("_", err).append(region, err);
+        if (U_SUCCESS(err)) {
```

Comment:
Error handling: Normally we would have this function take & check & update a `UErrorCode &` argument. If you don't want to do that, then we should at least try to return a non-broken string object, and not clobber an earlier failure like the later `err = U_ZERO_ERROR` currently does.

I suggest:
- pull the `result` declaration to near the top
- this line: `if (U_FAILURE(err)) { return result; }`
- don't reset `err` later (and you won't have to after bailing out on failure)

I think that should do. I believe that appending to result won't fail because the CharString has a stack buffer of 40 chars, so the script code will definitely fit.

Reply:
I agree.  I made the change you requested (or at least I think I did).